text/0000-named-arguments.md

```diff
@@ -0,0 +1,552 @@
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add named arguments to functions. Functions can have both positional and named arguments. In function calls, named arguments can be specified with their name to increase readability and maintainability.
+
+# Motivation
+[motivation]: #motivation
+
+Rust strives to be maintainable for large code bases. Named arguments would allow some function calls to be more explicit and thus easier to read and write.
+
+Let’s consider the following fictive examples:
+
+```rust
+http.get("https://www.rust-lang.org/", None);
+
+Window::new("Hello", 20, 50, 500, 250);
+
+free_fall(100.0, 0.0, 9.81);
+```
+
+Those are all functions that I could see written in Rust. Someone calling those functions has to remember the order in which the arguments appear. And at first sight, they can't really tell what the arguments stand for.
+
+A common workaround is to use longer function names, e.g. `get_with_timeout`, `new_with_bounds`, `free_fall_with_z0_and_v0_and_g`. However, function names that long are often considered ugly, and in the case of `new_with_bounds`, the order of arguments is still unclear.
+
+Now let’s consider calling the same functions with named arguments:
+
+```rust
+http.get("https://www.rust-lang.org/", .timeout = None);
+
+Window::new(.title = "Hello", .x = 20, .y = 50, .width = 500, .height = 250);
+
+free_fall(.z0 = 100.0, .v0 = 0.0, .g = 9.81);
+```
+
+It is now very clear what the arguments stand for, which makes maintaining such code easier.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Each function argument can be _named_ or _positional_. Positional arguments are identified by their argument position in function calls, named arguments are identified by their name:
+
+```rust
+positional_args(1, 42, "world", true);
+
+named_args(.x = 1, .y = 42, .hello = "world", .b = true);
+```
+
+## Function definition
+
+A named argument has an _argument name_ and an _argument pattern_. The argument name is part of the function's API, it's what is used _outside_ of the function. In contrast, the argument pattern is used _within_ the function.
+
+Usually, the argument pattern is the same as the argument name. The following function has a named argument `.x`, which is bound to a variable `x`:
+
+```rust
+fn identity(.x: i32) -> i32 { x }
+```
+
+It is also possible to specify a different pattern. This pattern can be used to rename or destructure the argument, for example:
+
+```rust
+fn add(.scalar a: i32, .to (b, c): (i32, i32)) -> i32 {
+    a + b + c
+}
+
+add(.scalar = 7, .to = (14, 21));
+```
+
+If a function has both positional and named arguments, the named arguments must come _after_ the positional arguments:
+
+```rust
+fn good(a: i32, .b: i32, .c: i32);
+
+fn bad(.a: i32, b: i32, .c: i32); // error!
+```
```

Comment:
I think the main use-case for mixed arguments is when you have a first argument which is clear, and then some options, like the `http::get(url, .some, .options)` example in the RFC. Arguably, in this case it is more readable for the first argument not to be named at the callsite, and then the API reflects this.

Reply:
The other use case is methods. The first parameter of a method (e.g. `self`, `&self`, `Rc<Self>`) must not be named according to this RFC.

This means that you can always write `foo.bar(.arg = 42)` instead of `Foo::foo(.receiver = foo, .arg = 42)`.

Reply:
Right, but methods are special-cased anyway. One could have a rule that arguments need to be all named or all unnamed, except for `self` which is always unnamed. One could even argue that having `self` as a name in case of `Foo::bar(.self = foo, .arg = 42)` actually improves readability. I would actually suggest allowing this, regardless of whether mixed args are allowed or not.

Reply:
Would the rule *all argument names need to be used* **xor** *no argument names need to be used* be okay?
The API implementation is intended for readability and I feel like you either want **full** or **no** readability (the latter requires checking source).  
The same argument goes for line 115-125 and I dont want to double-post.